Add DruidInputSource (replacement for IngestSegmentFirehose)

jon-wei opened the pull request:
Based on an initial patch by @jihoonson 

This PR adds `DruidInputSource` and associated classes, a new replacement for `IngestSegmentFirehose`. This input source always has a fixed input format (`DruidInputFormat`). The row reading class is `DruidSegmentReader` and the input entity is `DruidInputEntity`.

The new input source has the same parameters as the existing IngestSegmentFirehose, and has similar code.

Additionally:
- The PR adjusts the CompactionTask so that it uses DruidInputSource instead of the old firehose.
- Fixes an issue where the value of `inputSource.needsFormat()` was not being respected.
- The new input source avoids the issue described in #8448, which is present in the old firehose

For testing, this patch relies on the existing CompactionTask tests for checking functionality of the new input source. 

A test where an IngestSegmentFirehose is used for reindexing with a native batch task (non-parallel) has also been added to `CompactionRunTaskTest`. The existing `ITParallelIndexTaskTest` still uses a firehose instead of input source.

This PR has:
- [x] been self-reviewed.
- [x] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths.
- [x] added integration tests.
- [x] been tested in a test Druid cluster.

jon-wei commented:
Marking WIP, need to update the docs, and will add an integration test to `ITParallelIndexTaskTest` that uses the new input source.

vogievetsky commented:
Tested this (as a user) from the console (via https://github.com/apache/incubator-druid/pull/8828 ) and it works

jon-wei commented:
Added a doc entry for the new input source. We'll need a larger rework of the docs for the new input sources/input formats, but that should be handled in another PR.

Adjusted ITParallelIndexTaskTest so that it also runs reingestion using DruidInputSource

clintropolis commented on indexing-service/src/main/java/org/apache/druid/indexing/input/DruidInputSource.java:
```diff
@@ -0,0 +1,524 @@
+  }
+
+  @VisibleForTesting
+  private static List<String> getUniqueDimensions(
```
nit: Would this method and `getUniqueMetrics` maybe more appropriately live in `VersionedIntervalTimeline`?

jon-wei replied:
I moved these to VersionedIntervalTimeline, `IngestSegmentFirehoseFactory.testGetUniqueDimensionsAndMetrics` should be covering these now

ccaominh commented on indexing-service/src/main/java/org/apache/druid/indexing/input/DruidInputSource.java:
```diff
@@ -0,0 +1,524 @@
+  }
+
+  @VisibleForTesting
+  private static List<String> getUniqueMetrics(List<TimelineObjectHolder<String, DataSegment>> timelineSegments)
```
This method does not seem to be covered by any of the tests in `org.apache.druid.indexing`

jon-wei replied:
I moved these to VersionedIntervalTimeline, IngestSegmentFirehoseFactory.testGetUniqueDimensionsAndMetrics should be covering these now

ccaominh commented on indexing-service/src/main/java/org/apache/druid/indexing/input/DruidSegmentReader.java:
```diff
@@ -0,0 +1,225 @@
+                          for (int i = 0; i < valsSize; ++i) {
+                            dimVals.add(selector.lookupName(vals.get(i)));
+                          }
+                          theEvent.put(dim, dimVals);
```
This part does not seem to be covered by any of the tests in `org.apache.druid.indexing`

jon-wei replied:
I adjusted `CompactionRunTaskTest` so that there's a multivalued dimension (this case handles that), and added a row correctness check to `testRun` there

ccaominh commented on indexing-service/src/main/java/org/apache/druid/indexing/input/DruidInputSource.java:
```diff
@@ -0,0 +1,524 @@
+            // This segment won't fit in the current non-empty split, so this split is done.
+            splits.add(new InputSplit<>(currentSplit));
+            currentSplit = new ArrayList<>();
+            bytesInCurrentSplit = 0;
```
May be worth having a unit test for this method. Looks like this part is not covered by any of the tests in `org.apache.druid.indexing`.

jon-wei replied:
I added `testDruidInputSourceCreateSplitsWithIndividualSplits` to `CompactionTaskParallelRunTest`

jon-wei commented on docs/ingestion/native-batch.md:
```diff
@@ -890,3 +890,31 @@ This Firehose can be used to combine and merge data from a list of different Fir
+{
+    "type": "druid",
+    "dataSource": "wikipedia",
+    "interval": "2013-01-01/2013-01-02"
```
Added a second example with more explanation of what the example specs do

jon-wei commented on docs/ingestion/native-batch.md:
```diff
@@ -890,3 +890,31 @@ This Firehose can be used to combine and merge data from a list of different Fir
+|property|description|required?|
+|--------|-----------|---------|
+|type|This should be "druid".|yes|
+|dataSource|A String defining the data source to fetch rows from, very similar to a table in a relational database|yes|
```
Reworded this part and other parts of this table

jon-wei commented on indexing-service/src/main/java/org/apache/druid/indexing/input/DruidSegmentReader.java:
```diff
@@ -0,0 +1,225 @@
+        ),
+        source.getIntervalFilter()
+    );
+    final Sequence<Map<String, Object>> sequence = Sequences.concat(
```
Adjusted this to use the suggested structure

jihoonson replied:
Nice refactoring!

jon-wei commented on indexing-service/src/main/java/org/apache/druid/indexing/input/DruidSegmentInputEntity.java:
```diff
@@ -0,0 +1,99 @@
+      public void close()
+      {
+        if (!segmentFile.delete()) {
+          // log
```
Filled in a warning log message

jon-wei commented on indexing-service/src/main/java/org/apache/druid/indexing/input/DruidInputSource.java:
```diff
@@ -0,0 +1,524 @@
+    Interval lastInterval = null;
+    for (Interval interval : timeline.keySet()) {
+      if (lastInterval != null) {
+        if (interval.overlaps(lastInterval)) {
```
Collapsed the ifs

jon-wei commented on indexing-service/src/main/java/org/apache/druid/indexing/input/DruidInputSource.java:
```diff
@@ -0,0 +1,524 @@
+    // segments to olders.
+
+    // timelineSegments are sorted in order of interval
+    int[] index = {0};
```
I tried changing it to be the same as the dimensions version, but because there is only a presence check for the metrics version (no "exclusions" like with dimensions), the map.put becomes a style failure (tells me to use computeIfAbsent instead)

Made this final

jihoonson replied:
It could be `MutableInt`, but I think the original code is ok since it's just reusing the existing code.

jon-wei commented on indexing-service/src/main/java/org/apache/druid/indexing/input/DruidInputSource.java:
```diff
@@ -0,0 +1,524 @@
+  {
+    Preconditions.checkNotNull(interval);
+
+    // This call used to use the TaskActionClient, so for compatibility we use the same retry configuration
```
I left this for now

jihoonson replied:
I don't know why this has its own retry logic too.. I think it's worth to consider it for future refactoring.

ccaominh commented on docs/ingestion/native-batch.md:
```diff
@@ -890,3 +890,57 @@ This Firehose can be used to combine and merge data from a list of different Fir
+|dataSource|A String defining the Druid datasource to fetch rows from|yes|
+|interval|A String representing the ISO-8601 interval, which defines the time range to fetch the data over.|yes|
+|dimensions|A list of Strings containing the names of dimension columns to select from the Druid datasource. If the list is empty, no dimensions are returned. If null, all dimensions are returned. |no|
+|metrics|The list of Strings containg the names of metric columns to select. If the list is empty, no metrics are returned. If null, all metrics are returned.|no|
```
typo: containg -> containing

jon-wei replied:
Fixed spelling

jihoonson commented on core/src/main/java/org/apache/druid/timeline/VersionedIntervalTimeline.java:
```diff
@@ -753,6 +757,67 @@ private void remove(
   }
 
+  @VisibleForTesting
+  public static List<String> getUniqueDimensions(
```
The place of these methods seems a bit unintuitive to me. Their first parameter is a list of `TimelineObjectHolder`s because I think it's easier to use in `DruidInputSource` and `IngestSegmentFirehose`. But, in fact, a list of `DataSegments` would be enough. Is there any other place that needs to use these methods other than `DruidInputSource` or `IngestSegmentFirehose`? If not, I guess a new util class such as `CompactionUtils` or `DataSegments` might be better?

jon-wei replied:
I moved these to a new class `ReingestionTimelineUtils` and added some javadocs

jon-wei replied:
The existing comments within the method mention ordering based on recency of segments, so I kept the input as TimelineObjectHolders

clintropolis replied:
Hm, so I had suggested moving them to `VersionedIntervalTimeline` because they seem like generic timeline operations, not specific to re-ingesting segments. Is having a `ReingestionTimelineUtils` better than their original home if they are still going to live somewhere specific to re-ingesting segments?

jon-wei replied:
Hmm, my only view on this is it should be somewhere other than `IngestSegmentFirehoseFactory` since that's headed for eventual deprecation. 

It could go in `VersionedIntervalTimeline` although that class is a bit weird (it accepts generic types but also has methods that specifically use DataSegment). I feel like the utils class is fine for now

jihoonson replied:
My stance is similar to @jon-wei. It could be used for more general use cases, but I don't see them now. Maybe we can move them from `ReingestionTimelineUtils` to somewhere else if we need in the future.

jihoonson commented on indexing-service/src/main/java/org/apache/druid/indexing/input/DruidSegmentReader.java:
```diff
@@ -0,0 +1,277 @@
+    }
+
+    @Override
+    public void remove()
```
nit: the `Iterator` interface has the default implementation for `remove()` throwing `UnsupportedOperationException`.

jon-wei replied:
Removed the redundant override
